src/renate/utils/pytorch.py

```diff
@@ -156,3 +156,76 @@ def complementary_indices(num_outputs: int, valid_classes: Set[int]) -> List[int
         valid_classes: A set of integers of valid classes.
     """
     return [class_idx for class_idx in range(num_outputs) if class_idx not in valid_classes]
+
+
+class ConcatRandomSampler(BatchSampler):
+    """Sampler for sampling batches from ConcatDatasets.
+
+    Args:
+        dataset_lengths: The length for the different datasets.
+        batch_sizes: Batch sizes used for specific datasets.
+        complete_dataset_iteration: Provide an index to indicate over which dataset to fully
+            iterate. By default, stops whenever iteration is complete for any dataset.
+        generator: Generator used in sampling.
+        sampler: Lightning automatically passes a DistributedSamplerWrapper. Only used as an
+            indicator that we are in the distributed case.
+    """
+
+    def __init__(
+        self,
+        dataset_lengths: List[int],
+        batch_sizes: List[int],
+        complete_dataset_iteration: Optional[int] = None,
+        generator: Any = None,
+        sampler: Sampler = None,
+    ) -> None:
+        self.batch_sizes = batch_sizes
+        self.complete_dataset_iteration = complete_dataset_iteration
+        self.subset_samplers = []
+        data_start_idx = 0
+        num_batches = []
+        rank = 0 if sampler is None else sampler.rank
+        num_replicas = 1 if sampler is None else sampler.num_replicas
+        for dataset_length, batch_size in zip(dataset_lengths, batch_sizes):
+            data_end_idx = data_start_idx + dataset_length
+            start_idx = data_start_idx + round(dataset_length / num_replicas * rank)
+            end_idx = data_start_idx + round(dataset_length / num_replicas * (rank + 1))
+            subset_sampler = BatchSampler(
+                SubsetRandomSampler(list(range(start_idx, end_idx)), generator),
+                batch_size,
+                True,
+            )
+            self.subset_samplers.append(subset_sampler)
+            num_batches.append((end_idx - start_idx) // batch_size)
+            data_start_idx = data_end_idx
+        self.length = (
+            min(num_batches)
+            if complete_dataset_iteration is None
```

Comment:
Why not `self.complete_dataset_iteration` here?

Reply:
done